src/xmpp/xmpp-im/xmpp_mamtask.cpp

```diff
@@ -0,0 +1,255 @@
+ * along with this library.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "xmpp_mamtask.h"
+
+using namespace XMLHelper;
+using namespace XMPP;
+
+class MAMTask::Private {
+public:
+    int     mamPageSize; // TODO: this is the max page size for MAM request. Should be made into a config option in Psi+
+    int     mamMaxMessages; // maximum mam pages total, also should be config. zero means unlimited
+    int     messagesFetched;
+    bool    flipPages;
+    bool    backwards;
+    bool    allowMUCArchives;
+    bool    metadataFetched;
+    Jid     j;
+    QString firstID;
+    QString lastID;
+    QString lastArchiveID;
+    QString from_id;
+    QString to_id;
+    QDateTime          from;
+    QDateTime          to;
+    QList<QDomElement> archive;
+
+    void  getPage(MAMTask *t);
+    void  getArchiveMetadata(MAMTask *t);
+    XData makeMAMFilter();
+};
+
+
+MAMTask::MAMTask(Task *parent) : Task(parent) { d = new Private; }
+MAMTask::MAMTask(const MAMTask& x) : Task(x.parent()) { d = x.d; }
+MAMTask::~MAMTask() { delete d; }
+
+const QList<QDomElement> &MAMTask::archive() const { return d->archive; }
+
+XData MAMTask::Private::makeMAMFilter()
+{
+    XData::FieldList fl;
+
+    XData::Field with;
+    with.setType(XData::Field::Field_JidSingle);
+    with.setVar(QLatin1String("with"));
+    with.setValue(QStringList(j.full()));
+    fl.append(with);
+
+    XData::Field includeGroupchat;
+    includeGroupchat.setType(XData::Field::Field_Boolean);
+    includeGroupchat.setVar(QLatin1String("include-groupchat"));
+    includeGroupchat.setValue(QStringList(QLatin1String(allowMUCArchives ? "true" : "false")));
+    fl.append(includeGroupchat);
+
+    if (from.isValid()) {
+        XData::Field start;
+        start.setType(XData::Field::Field_TextSingle);
+        start.setVar(QLatin1String("start"));
+        from.setTimeSpec(Qt::UTC);
+        start.setValue(QStringList(from.toString()));
+        fl.append(start);
+    }
+
+    if (to.isValid()) {
+        XData::Field end;
+        end.setType(XData::Field::Field_TextSingle);
+        end.setVar(QLatin1String("end"));
+        to.setTimeSpec(Qt::UTC);
+        end.setValue(QStringList(to.toString()));
+        fl.append(end);
+    }
+
+    if (!from_id.isNull()) {
+        XData::Field start_id;
+        start_id.setType(XData::Field::Field_TextSingle);
+        start_id.setVar(QLatin1String("after-id"));
+        start_id.setValue(QStringList(from_id));
+        fl.append(start_id);
+    }
+
+    if (!to_id.isNull()) {
+        XData::Field end_id;
+        end_id.setType(XData::Field::Field_TextSingle);
+        end_id.setVar(QLatin1String("before-id"));
+        end_id.setValue(QStringList(to_id));
+        fl.append(end_id);
+    }
+
+    XData x;
+    x.setType(XData::Data_Submit);
+    x.setFields(fl);
+    x.setRegistrarType(XMPP_MAM_NAMESPACE);
+
+    return x;
+}
+
+void MAMTask::Private::getPage(MAMTask *t)
```

Comment:
better pass the task once to the Private's constructor and remember there. You `q` member for that, which is sort of standard